daemon/src/position_metrics.rs

```diff
@@ -0,0 +1,344 @@
+                "Total quantity of positions on ItchySats.",
+                &[POSITION_LABEL, STATUS_LABEL]
+            )
+            .unwrap()
+        });
+
+    static POSITION_AMOUNT_GAUGE: conquer_once::Lazy<prometheus::IntGaugeVec> =
+        conquer_once::Lazy::new(|| {
+            prometheus::register_int_gauge_vec!(
+                "positions_total",
+                "Total number of positions on ItchySats.",
+                &[POSITION_LABEL, STATUS_LABEL]
+            )
+            .unwrap()
+        });
+
+    pub fn update_position_metrics(cfds: &[Cfd]) {
+        set_position_metrics(cfds.iter().filter(|cfd| cfd.is_open), STATUS_OPEN_LABEL);
+        set_position_metrics(cfds.iter().filter(|cfd| cfd.is_closed), STATUS_CLOSED_LABEL);
+        set_position_metrics(cfds.iter().filter(|cfd| cfd.is_failed), STATUS_FAILED_LABEL);
+        set_position_metrics(
+            cfds.iter().filter(|cfd| cfd.is_refunded),
+            STATUS_REFUNDED_LABEL,
+        );
+
+        set_position_metrics(
+            cfds.iter().filter(|cfd| {
+                let unknown_state = !(cfd.is_open ^ cfd.is_closed ^ cfd.is_refunded);
+                if unknown_state {
+                    tracing::error!(
+                        is_open = cfd.is_open,
+                        is_closed = cfd.is_closed,
+                        is_refunded = cfd.is_refunded,
+                        order_id = %cfd.id,
+                        "CFD is in weird state"
+                    );
+                }
+                unknown_state
+            }),
+            STATUS_UNKNOWN_LABEL,
+        );
+    }
+
+    fn set_position_metrics(cfds: Filter<Iter<Cfd>, fn(&&Cfd) -> bool>, status: &str) {
+        let (long, short): (Vec<_>, Vec<_>) = cfds.partition(|cfd| cfd.position == Position::Long);
+
+        POSITION_QUANTITY_GAUGE
+            .with(&HashMap::from([
+                (POSITION_LABEL, POSITION_LONG_LABEL),
+                (STATUS_LABEL, status),
+            ]))
+            .set(
+                sum_amounts(&long)
+                    .into_decimal()
+                    .to_f64()
+                    .unwrap_or_default(),
+            );
+        POSITION_AMOUNT_GAUGE
+            .with(&HashMap::from([
+                (POSITION_LABEL, POSITION_LONG_LABEL),
+                (STATUS_LABEL, status),
+            ]))
+            .set(long.len() as i64);
+
+        POSITION_QUANTITY_GAUGE
+            .with(&HashMap::from([
+                (POSITION_LABEL, POSITION_SHORT_LABEL),
+                (STATUS_LABEL, status),
+            ]))
+            .set(
+                sum_amounts(&short)
+                    .into_decimal()
+                    .to_f64()
+                    .unwrap_or_default(),
+            );
+        POSITION_AMOUNT_GAUGE
+            .with(&HashMap::from([
+                (POSITION_LABEL, POSITION_SHORT_LABEL),
+                (STATUS_LABEL, status),
+            ]))
+            .set(short.len() as i64);
+    }
```

Comment:
extracting the repetition made it much easier to follow;

```suggestion
 
        set_metrics_for(POSITION_LONG_LABEL, status, &long);
        set_metrics_for(POSITION_SHORT_LABEL, status, &short);
    }

    fn set_metrics_for(position_label: &str, status: &str, position: &[&Cfd]) {
        POSITION_QUANTITY_GAUGE
            .with(&HashMap::from([
                (POSITION_LABEL, position_label),
                (STATUS_LABEL, status),
            ]))
            .set(
                sum_amounts(position)
                    .into_decimal()
                    .to_f64()
                    .unwrap_or_default(),
            );
        POSITION_AMOUNT_GAUGE
            .with(&HashMap::from([
                (POSITION_LABEL, position_label),
                (STATUS_LABEL, status),
            ]))
            .set(position.len() as i64);
    }
```